### DISCUSSION TITLE
How can I capture key-pressed events while typing in `LineEdit`?

### DISCUSSION THREAD

**Discussion by Horbin-Magician** [slint-ui/slint#4063] [Q&A]
The `LineEdit` will capture all related key-pressed event. But sometimes I need to handle some key-pressed events such as direction keys.

For example, when entering the search content, press the arrow key to select the corresponding item:

https://github.com/slint-ui/slint/assets/46142827/58f94b4d-5a9a-4c4a-8b1b-8517887ef551

I am currently implementing it with the key-released event. But that way, I can't implement the long press event, i. e., keep pressing the down key, and the selected  item keeps moving down.

Is there any way to do that?

**Comment by ogoffart** [comment]
Are you saying that you can intercept the key-released but not the key-press?
Is that because the LineEdit intercept the key-press but not the key-release?

That's indeed a strange behavior, the TextInput (part of the LineEdit)  should intercept both the same way. But that wouldn't help you.

I wonder how we could make that work.
We would need a focus scope that intercept events BEFORE for LineEdit.  
@tronical any idea on how this could be represented in .slint?

**Reply by Horbin-Magician** [reply]
> Are you saying that you can intercept the key-released but not the key-press?

Yes, but I still can intercept some key-press event such as `shift`, `ctrl`.

> Is that because the LineEdit intercept the key-press but not the key-release?

Yes, only the key event which is useless for LineEdit can be intercept by external focus scope component.

I think this is a reasonable way to deal with event. But it is not flexible enough.

I have an underdeveloped idea:
Maybe we need a way to establish event transmission between components.
For example, if the LineEdit can copy part of its event to FocusScope, It seems that many features can be easily implemented.

**Reply by tronical** [reply]
It sounds like that we may need to concept of an event filter for FocusScope (or focusable objects in general). Like `FocusScope { filter-from: name-of-element; ... }`?

**Reply by Horbin-Magician** [reply]
> It sounds like that we may need to concept of an event filter for FocusScope (or focusable objects in general). Like FocusScope { filter-from: name-of-element; ... }?

Yes, I think so.

**Reply by ogoffart** [reply]
I don't think we need a `filter-from`  we should just filter anything that is a children of the scope.

Anyway, there is now an issue for this: https://github.com/slint-ui/slint/issues/4337

**Comment by Horbin-Magician** [comment]
I have found this: https://github.com/slint-ui/slint/issues/145
This is also one way. Is this a change in the plan? @ogoffart @tronical
